### PR TITLE
#9295: Yolov4 relu fused convs integration

### PR DESCRIPTION
**Pull request by ankitmcw**
This pr contains yolov4 integration with relu fused ttnn convs.

Perf results:

Downsample1 submodule: [ops_perf_results_DownSample1_2024_06_07_10_58_05.csv](https://github.com/user-attachments/files/15741556/ops_perf_results_DownSample1_2024_06_07_10_58_05.csv)

Downsample2 submodule: [ops_perf_results_DownSample2_2024_06_07_11_14_26.csv](https://github.com/user-attachments/files/15741559/ops_perf_results_DownSample2_2024_06_07_11_14_26.csv)

Downsample3 submodule: [ops_perf_results_DownSample3_2024_06_07_11_20_10.csv](https://github.com/user-attachments/files/15741562/ops_perf_results_DownSample3_2024_06_07_11_20_10.csv)

Downsample4 submodule: [ops_perf_results_DownSample4_2024_06_07_11_25_07.csv](https://github.com/user-attachments/files/15741563/ops_perf_results_DownSample4_2024_06_07_11_25_07.csv)

Downsample5 submodule: [ops_perf_results_DownSample5_2024_06_07_11_29_08.csv](https://github.com/user-attachments/files/15741565/ops_perf_results_DownSample5_2024_06_07_11_29_08.csv)

Head submodule: [ops_perf_results_Head_2024_06_11_09_11_24.csv](https://github.com/user-attachments/files/15786324/ops_perf_results_Head_2024_06_11_09_11_24.csv)

Neck submodule: [ops_perf_results_Neck_2024_06_11_10_02_54.csv](https://github.com/user-attachments/files/15786325/ops_perf_results_Neck_2024_06_11_10_02_54.csv)

Yolov4 model: [ops_perf_results_Yolov4_2024_06_11_10_05_09.csv](https://github.com/user-attachments/files/15786326/ops_perf_results_Yolov4_2024_06_11_10_05_09.csv)

Note: 
For replacing only Mish with relu in fused conv, retain LeakyRelu as is
1. For Downsample1-5 getting 0.99 pcc.
2. PCC neck
       out_1 = 0.9383100471190877
       out_2 = 0.9391142122379718
       out_3 = 0.9299983109764207
3. PCC head
       out_1 = 0.9858622715837794
       out_2 = 0.9707051554758256
       out_3 = 0.7837755285688921
4. PCC model
       out_1 = 0.8884796846649758
       out_2 = 0.8951969069533124
       out_3 = 0.02981400787745157

### CODE REVIEW

File: models/experimental/functional_yolov4/tt/ttnn_resblock.py [hunk 1]
```diff
@@ -17,10 +19,15 @@ def __init__(self, parameters, nblocks, shortcut) -> None:
...
-        input_tensor = input_tensor.to(device, self.module_list[0][0].conv.input_sharded_memory_config)
+        input_tensor = tt_lib.tensor.sharded_to_interleaved(input_tensor, ttnn.L1_MEMORY_CONFIG)
+        input_tensor = ttnn.to_layout(input_tensor, layout=ttnn.TILE_LAYOUT)
         for i in range(self.nblocks):
             output_tensor_h = input_tensor
+            output_tensor_h = output_tensor_h.to(device, self.module_list[i][0].conv.input_sharded_memory_config)
             output_tensor_1 = self.module_list[i][0](output_tensor_h)
             output_tensor_h = self.module_list[i][1](output_tensor_1)
-            input_tensor = ttnn.add(input_tensor, output_tensor_h) if self.shortcut else output_tensor_h
+            output_tensor_h = tt_lib.tensor.sharded_to_interleaved(output_tensor_h, ttnn.L1_MEMORY_CONFIG)
```
Comment (dvartaniansTT) [inline models/experimental/functional_yolov4/tt/ttnn_resblock.py:29]: we can remove this

File: models/experimental/functional_yolov4/tt/ttnn_resblock.py [hunk 2]
```diff
@@ -17,10 +19,15 @@ def __init__(self, parameters, nblocks, shortcut) -> None:
             self.module_list.append(resblock_one)
 
     def __call__(self, device, input_tensor):
-        input_tensor = input_tensor.to(device, self.module_list[0][0].conv.input_sharded_memory_config)
+        input_tensor = tt_lib.tensor.sharded_to_interleaved(input_tensor, ttnn.L1_MEMORY_CONFIG)
```
Comment (dvartaniansTT) [inline models/experimental/functional_yolov4/tt/ttnn_resblock.py:22]: remove this

File: models/experimental/functional_yolov4/tt/ttnn_downsample1.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        self.c4 = parameters.c4
+        self.c5 = parameters.c5
+        self.c6 = parameters.c6
+        self.c7 = parameters.c7
+        self.c8 = parameters.c8
+
+    def __call__(self, device, input_tensor):
+        input_tensor = input_tensor.to(device, self.c1.conv.input_sharded_memory_config)
+
+        output_tensor = self.c1(input_tensor)
```
Comment (dvartaniansTT) [inline models/experimental/functional_yolov4/tt/ttnn_downsample1.py:31]: call the new conv api and provide reshard_if_not_optimal==True. you must provide the shard scheme height_sharding= True

File: tests/ttnn/integration_tests/yolov4/custom_preprocessor_d1.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+from models.experimental.functional_yolov4.reference.downsample1 import DownSample1
+from models.experimental.functional_yolov4.tt.ttnn_downsample1 import TtDownSample1
+
+import ttnn
+import tt_lib
+
+
+def update_ttnn_module_args(ttnn_module_args):
+    ttnn_module_args["use_1d_systolic_array"] = ttnn_module_args.in_channels < 256
```
Comment (dvartaniansTT) [inline tests/ttnn/integration_tests/yolov4/custom_preprocessor_d1.py:16]: height sharding is True

File: tests/ttnn/integration_tests/yolov4/custom_preprocessor_d1.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+
+def custom_preprocessor(device, model, name, ttnn_module_args):
+    parameters = {}
+    if isinstance(model, DownSample1):
+        ttnn_module_args.c1["math_fidelity"] = ttnn.MathFidelity.LoFi
```
Comment (dvartaniansTT) [inline tests/ttnn/integration_tests/yolov4/custom_preprocessor_d1.py:22]: struct Conv2dConfig {
MathFidelity math_fidelity = MathFidelity::HiFi4;
DataType dtype = DataType::BFLOAT16;
DataType weights_dtype = DataType::BFLOAT16;
bool math_approx_mode_enabled = true;
bool fp32_dest_acc_enabled = false;
bool packer_l1_accum_enabled = false;
string activation = "";
uint32_t input_channels_alignment = 32;
bool deallocate_activation = false;
bool reallocate_halo_output = false;
uint32_t act_block_h_override = 0;
bool reshard_if_not_optimal = false; // if true, override_sharding_config should not be set to true
bool override_sharding_config = false; // if true, reshard_if_not_optimal should not be set to true
bool height_sharding = true; // used only if override_sharding_config is true
std::optional<CoreRangeSet> core_grid = std::nullopt; // used only if override_sharding_config is true
bool transpose_shards = true; // used only if override_sharding_config is true and if height sharding is false
Layout output_layout = Layout::TILE;

Comment (dvartaniansTT) [reply tests/ttnn/integration_tests/yolov4/custom_preprocessor_d1.py:22]: uint32_t input_channels_alignment = 16. -> shallow conv

Comment (dvartaniansTT) [reply tests/ttnn/integration_tests/yolov4/custom_preprocessor_d1.py:22]: uint32_t act_block_h_override = 0 -> None

**Review by TT-BrianLiu — APPROVED** [review]

**Pull request closed without merging**